Cache-Hash-Data-Structure.adoc

```diff
@@ -0,0 +1,174 @@
+= Hash Data Structure
+
+Provide a Hash data structure to Infinispan users. The Hash is a flexible identifiable grouping of keys mapping to values.
+In this context, _flexible_ means it can have arbitrary mappings without needing a schema beforehand; _identifiable group_
+means the Hash structure receives a unique name and retrieves the complete key-value mapping.
+
+The Hash provides a general approach for users to represent their data structures using Infinispan. For example, by
+employing this, users can avoid using key prefixes to create an object in Cache.
+
+== Public API
+
+The work is for the new API only. The Hash data structure has the format of:
+
+[source,java]
+.Hash.java
+----
+public interface Hash<K, V> { // <1>
+
+   String getName();
+
+   V get(K key);
+
+   V put(K key, V value);
+
+   V remove(K key);
+
+   boolean containsKey(K key);
+
+   Set<K> keySet();
+
+   Collection<V> values();
+
+   Set<Map.Entry<K, V>> entrySet();
+
+   void clear();
+
+   int size();
+}
+----
+
+<1> The Hash data structure is a generic interface over `K` and `V`. Should we extend `Map<K, V>`?
+This way we could provide a wide range of operations over the Hash.
+
+A `Hash<?, ?>` structure is local to a node. Meaning that operations applied to the local copy must be submitted
+through the public API before taking effect. After that, the `Hash<?, ?>` is distributed/replicated according to
+the configuration.
```

Comment:
How is the intended use case?
The user operates on an Hash object and then uses an HashCache obj to submit the changes?

Reply:
I would say that individual operations are applied immediately to the underlying storage (remote cache). When using tx/batching, these would be collected and sent in a single operation.

Reply:
> The user operates on an Hash object and then uses an HashCache obj to submit the changes?

Yeah, I was thinking of something of that sort. I'll update the documentation to include single operations without transferring the whole object. For multiple operations, I was thinking of keeping track of the local changes, and once the user submits the changes to the server, we merge the local Hash with the server's Hash.

I'll update the docs.